libutils/regex.c

```diff
@@ -172,12 +197,43 @@ Seq *StringMatchCaptures(const char *regex, const char *str)
         return NULL;
     }
 
-    Seq *ret = SeqNew(captures + 1, free);
+    Seq *ret = SeqNew(captures + 1, BufferDestroy);
     for (int i = 0; i <= captures; ++i)
     {
-        SeqAppend(ret, xstrndup(str + ovector[2*i],
-                                ovector[2*i + 1] - ovector[2 * i]));
+        Buffer *capture = NULL;
+
+        if (have_named_captures)
+        {
+            // The overhead of doing a nested name scan is negligible.
+            tabptr = name_table;
+            for (int namepos = 0; namepos < namecount; namepos++)
+            {
+                int n = (tabptr[0] << 8) | tabptr[1];
+                if (n == i) // We found the position
+                {
+                    capture = BufferNewFrom(tabptr + 2, name_entry_size - 3);
+                    break;
+                }
+                tabptr += name_entry_size;
+            }
+        }
+
+        if (return_names)
+        {
```

Comment:
This is guaranteed to be true if ``have_named_captures`` is true (it depends on it), so just move this block into the previous block.